packages/appium/lib/schema/schema.js

```diff
@@ -0,0 +1,571 @@
+
+    /**
+     *
+     * @param {SchemaObject} schema
+     * @param {ExtensionType} [extType]
+     * @param {string} [extName]
+     */
+    const addArgSpecs = (schema, extType, extName) => {
+      _.forEach(schema, (propSchema, propName) => {
+        const argSpec = ArgSpec.create(propName, {
+          dest: propSchema.appiumCliDest,
+          defaultValue: propSchema.default,
+          extType,
+          extName,
+        });
+        const {arg} = argSpec;
+        this._argSpecs.set(arg, argSpec);
+      });
+    };
+
+    addArgSpecs(
+      _.omit(baseSchema.properties.server.properties, [
+        DRIVER_TYPE,
+        PLUGIN_TYPE,
+      ]),
+    );
+
+    /**
+     * @type {Record<string,StrictSchemaObject>}
+     */
+    const finalizedSchemas = {};
+
+    const finalSchema = _.reduce(
+      this._registeredSchemas,
+      /**
+       * @param {typeof baseSchema} baseSchema
+       * @param {Map<string,SchemaObject>} extensionSchemas
+       * @param {ExtensionType} extType
+       */
+      (baseSchema, extensionSchemas, extType) => {
+        extensionSchemas.forEach((schema, extName) => {
+          const $ref = ArgSpec.toSchemaBaseRef(extType, extName);
+          schema.$id = $ref;
+          schema.additionalProperties = false; // this makes `schema` become a `StrictSchemaObject`
+          baseSchema.properties.server.properties[extType].properties[extName] =
+            {$ref, $comment: extName};
+          ajv.validateSchema(schema, true);
+          addArgSpecs(schema.properties, extType, extName);
+          ajv.addSchema(schema, $ref);
+          finalizedSchemas[$ref] = /** @type {StrictSchemaObject} */(schema);
+        });
+        return baseSchema;
+      },
+      baseSchema,
+    );
+
+    ajv.addSchema(finalSchema, APPIUM_CONFIG_SCHEMA_ID);
+    finalizedSchemas[APPIUM_CONFIG_SCHEMA_ID] = finalSchema;
+    ajv.validateSchema(finalSchema, true);
+
+    this._finalizedSchemas = finalizedSchemas;
+    return Object.freeze(finalizedSchemas);
+  }
+
+  /**
+   * Configures and creates an Ajv instance.
+   * @private
+   * @returns {Ajv}
+   */
+  static _instantiateAjv () {
+    const ajv = addFormats(
+      new Ajv({
+        // without this not much validation actually happens
+        allErrors: true,
+      }),
+    );
+
+    // add custom keywords to ajv. see schema-keywords.js
+    _.forEach(keywords, (keyword) => {
+      ajv.addKeyword(keyword);
+    });
+
+    return ajv;
+  }
+
+  /**
+   * Resets this instance to its original state.
+   *
+   * - Removes all added schemas from the `Ajv` instance
+   * - Resets the map of {@link ArgSpec ArgSpecs}
+   * - Resets the map of registered schemas
+   * - Sets the {@link AppiumSchema._finalized _finalized} flag to `false`
+   *
+   * If you need to call {@link AppiumSchema.finalize} again, you'll want to call this first.
+   * @public
+   * @returns {void}
+   */
+  reset () {
```

Comment:
why do we need to be able to reset the schema? we're not reading anything on the fly after the server is loaded right?

Reply:
because `AppiumSchema` is a singleton, the tests need it.  but otherwise, it may be useful for something consuming Appium.  I don't think it needs to be marked explicitly as a public API.

Reply:
makes sense!